pkg/nano/duration.go

```diff
@@ -0,0 +1,173 @@
+package nano
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Duration int64
+
+type unit struct {
+	name string
+	size int64
+}
+
+var units = []unit{
+	{"y", int64(Year)},
+	{"d", int64(Day)},
+	{"h", int64(Hour)},
+	{"m", int64(time.Minute)},
+}
+
+const minDur = "-292y171d23h47m16.854775808s"
+
+func (d Duration) String() string {
+	if int64(d) == math.MinInt64 {
+		return minDur
+	}
+	if d == 0 {
+		return "0s"
+	}
+	var b strings.Builder
+	ns := int64(d)
+	if ns < 0 {
+		ns = -ns
+		b.WriteByte('-')
+	}
+	for _, u := range units {
+		if ns >= u.size {
+			nunit := ns / u.size
+			ns -= nunit * u.size
+			b.WriteString(strconv.FormatInt(nunit, 10))
+			b.WriteString(u.name)
+			if ns == 0 {
+				return b.String()
+			}
+		}
+	}
+	switch {
+	case ns%1_000_000_000 == 0:
+		b.WriteString(strconv.FormatInt(ns/1_000_000_000, 10))
+		b.WriteString("s")
+	case ns > 1_000_000_000:
+		writeFixedPoint(&b, ns, 1_000_000_000)
+		b.WriteString("s")
+	case ns%1_000_000 == 0:
+		b.WriteString(strconv.FormatInt(ns/1_000_000, 10))
+		b.WriteString("ms")
+	case ns > 1_000_000:
+		writeFixedPoint(&b, ns, 1_000_000)
+		b.WriteString("ms")
+	case ns%1_000 == 0:
+		b.WriteString(strconv.FormatInt(ns/1_000, 10))
+		b.WriteString("us")
+	case ns > 1_000:
+		writeFixedPoint(&b, ns, 1_000)
+		b.WriteString("us")
+	default:
+		b.WriteString(strconv.FormatInt(ns, 10))
+		b.WriteString("ns")
+	}
+	return b.String()
+}
+
+func writeFixedPoint(b *strings.Builder, ns, scale int64) {
+	v := ns / scale
+	ns -= v * scale
+	b.WriteString(strconv.FormatInt(v, 10))
+	b.WriteByte('.')
+	scale /= 10
+	for ns > 0 {
+		digit := ns / scale
+		b.WriteByte('0' + byte(digit))
+		ns -= digit * scale
+		scale /= 10
+	}
+}
+
+func (d Duration) MarshalJSON() ([]byte, error) {
+	s := d.String()
+	return json.Marshal(&s)
+}
+
+func (d *Duration) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	v, err := ParseDuration(s)
+	if err != nil {
+		return err
+	}
+	*d = v
+	return nil
+}
+
+var parseRE = regexp.MustCompile("([.0-9]+)(ns|us|ms|s|m|h|d|w|y)")
+var syntaxRE = regexp.MustCompile("^([.0-9]+(ns|us|ms|s|m|h|d|w|y))+$")
+
+var scale = map[string]Duration{
+	"ns": Nanosecond,
+	"us": Microsecond,
+	"ms": Millisecond,
+	"s":  Second,
+	"m":  Minute,
+	"h":  Hour,
+	"d":  Day,
+	"w":  Week,
+	"y":  Year,
```

Comment:
Can a year duration be specified given its variable length? I noticed you don't have a test for it...

Reply:
This is a limitation of the syntax and it's worse for months so we don't have a month unit at all.  Unless you know the years over which the duration falls (and even the time zone!), you can't interpret the "y" unit as the precise number of years.  That said, the encoding is precise in that "y" is an abstract unit exactly equal to 365 days... it just doesn't reflect date differences precisely.  For this, we'd need a moments type library accessible via function calls and time difference calculations.  BTW, this is how Prometheus works so we are in good company.

Reply:
That said I can of course add a boomerang for "y".

Reply:
I will clarify the ZSON spec on this point.